docs/source/plugins.rst

```diff
@@ -0,0 +1,194 @@
+
+Launch the MLflow UI:
+
+.. code-block:: bash
+
+  cd ..
+  mlflow server --backend-store-uri ./mlflow/mlruns
+
+
+View results at http://localhost:5000
+
+
+Writing Your Own MLflow Plugin(s)
+---------------------------------
+
+Defining a Plugin
+~~~~~~~~~~~~~~~~~
+You define an MLflow plugin as a standalone Python package which can then be distributed for
+installation via PyPI or conda. See https://github.com/mlflow/mlflow/tree/master/tests/resources/mlflow-test-plugin for an
+example package that implements all currently-supported plugin types.
+
+In particular, the example package contains a ``setup.py`` that declares a number of
+`entry points <https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins>`_:
+
+.. code-block:: python
+
+    setup(
+        name="mflow-test-plugin",
+        # Require MLflow as a dependency of the plugin, so that plugin users can simply install
+        # the plugin and then immediately use it with MLflow
+        install_requires=["mlflow"],
+        ...
+        entry_points={
+            # Define a tracking AbstractStore plugin for tracking URIs with scheme 'file-plugin'
+            "mlflow.tracking_store": "file-plugin=mlflow_test_plugin:PluginFileStore",
+            # Define a ArtifactRepository plugin for artifact URIs with scheme 'file-plugin'
+            "mlflow.artifact_repository":
+                "file-plugin=mlflow_test_plugin:PluginLocalArtifactRepository",
+            # Define a RunContextProvider plugin. The entry point name for run context providers
+            # is not used, and so is set to the string "unused" here
+            "mlflow.run_context_provider": "unused=mlflow_test_plugin:PluginRunContextProvider",
+            # Define a model-registry AbstractStore plugin for tracking URIs with scheme 'file-plugin'
+            "mlflow.model_registry_store":
+                "file-plugin=mlflow_test_plugin:PluginRegistrySqlAlchemyStore",
+        },
+    )
+
+Each element of this ``entry_points`` dictionary specifies a single plugin. You
+can choose to implement one or more plugin types in your package, and need not implement them all.
+The type of plugin defined by each entry point and its corresponding reference implementation in
+MLflow are described below. You can work from the reference implementations when writing your own
+plugin:
+
+.. list-table::
+   :widths: 10 10 80 10
+   :header-rows: 1
+
+   * - Description
+     - Entry-point group
+     - Entry-point name and value
+     - Reference Implementation
+   * - Plugins for overriding definitions of tracking APIs like ``mlflow.log_metric``, ``mlflow.start_run`` for a specific
+       tracking URI scheme.
+     - mlflow.tracking_store
+     - The entry point value (e.g. ``mlflow_test_plugin:PluginFileStore``) specifies a custom subclass of
+       `mlflow.tracking.store.AbstractStore <https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/abstract_store.py#L8>`_
```

Comment:
@smurching Some of the links in this table refer to specific lines in files. Others simply refer to the file. If we're going to refer to specific line numbers, I think we may want to consider using permalinks in case the files change. Either way, I think we should be consistent (permalink to lines for class / function definitions everywhere - my preferred approach - or link to files everywhere).

Reply:
Sounds good, will switch to permalinks - linking to specific lines seems helpful, & I don't think using permalinks will require us to spend a ton of time updating the docs (e.g. seems unlikely - and actually not possible for stable APIs like tracking & artifact repositories - that the core API/classes will change dramatically, even if we add/remove code from the files in master)